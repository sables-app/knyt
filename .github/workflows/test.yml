name: Test

on:
  push:
    branches: [main]
  pull_request:
    branches: [main]

jobs:
  test:
    name: Test Suite
    runs-on: ubuntu-latest

    steps:
      - name: Checkout code
        uses: actions/checkout@v4

      - name: Setup Bun
        uses: oven-sh/setup-bun@v2
        with:
          bun-version: "latest"

      - name: Install dependencies
        run: bun install

      - name: Build packages
        run: |
          bun run clean
          bun run build

      - name: Run test suite
        run: |
          bun run check
          bun run test
