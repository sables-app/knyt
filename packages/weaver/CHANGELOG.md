# @knyt/weaver

## 0.4.0

### Minor Changes

- 3f1f128: - Update all relative import paths to be explicit by add file extensions.
  - This improves compatibility with tools that require explicit file extensions in import statements; making the project more runtime agnostic.
  - Update Bun engine requirement to "^1.3.0" in all package.json files.
    - Run `bun upgrade` to update the latest version of Bun.
    - This is a `minor` change because we're still in the `0.x` version range; also the public API of the package remains unchanged.

### Patch Changes

- Updated dependencies [3f1f128]
  - @knyt/artisan@0.4.0
  - @knyt/html-type@0.4.0

## 0.3.0

### Patch Changes

- 3d7b43c: Correct handling of empty strings in development-only assertions during updates
- eee0d94: Fixed broken event listener comparison.
- Updated dependencies [faa654e]
  - @knyt/artisan@0.3.0

## 0.2.0

### Patch Changes

- Updated dependencies [3bcb930]
  - @knyt/artisan@0.2.0

## 0.1.0

Initial release.
