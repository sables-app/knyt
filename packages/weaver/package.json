{
  "name": "@knyt/weaver",
  "version": "0.1.0-alpha",
  "description": "A light-weight DOM Renderer supports Web components standard & TypeScript language",
  "author": "Morris Allison III (https://morris.xyz)",
  "license": "BSD-3-Clause",
  "bugs": "https://github.com/sables-app/knyt/issues",
  "homepage": "https://github.com/sables-app/knyt",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/sables-app/knyt.git",
    "directory": "packages/weaver"
  },
  "type": "module",
  "module": "./dist/build/main.js",
  "main": "./dist/build/main.js",
  "exports": {
    ".": "./dist/build/main.js",
    "./jsx-runtime": "./dist/build/jsx.js",
    "./jsx-dev-runtime": "./dist/build/jsx.js",
    "./package.json": "./package.json"
  },
  "dependencies": {
    "@knyt/artisan": "^0.1.0-alpha",
    "@knyt/html-type": "^0.1.0-alpha"
  },
  "files": [
    "dist",
    "!__tests__",
    "!*.tsbuildinfo"
  ]
}
