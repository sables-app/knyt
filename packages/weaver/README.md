<div align="center">

[![Knyt](./docs/banner.svg)](https://knyt.dev/s/weaver)

ðŸ§µ Declarative DOM renderer with 1:1 native browser APIs. Strictly separates HTML, DOM, and SVG with fully type-safe, explicit interfaces for props, children and refs.

<small>

This package is part of [Knyt](https://knyt.dev/), a toolkit designed to simplify modern web development.

</small>

[![npm](https://img.shields.io/npm/v/@knyt/weaver?style=flat-square&labelColor=444)](https://www.npmjs.com/package/@knyt/weaver)
[![GitHub](https://img.shields.io/badge/Source_Code-black?style=flat-square&label=GitHub&labelColor=444)](https://github.com/sables-app/knyt/tree/main/packages/weaver)
[![License](https://img.shields.io/badge/License-BSD_3_Clause-blue?style=flat-square&labelColor=444)](https://github.com/sables-app/knyt/blob/main/LICENSE)
<br />
![Built with TypeScript](https://img.shields.io/badge/Built%20with-TypeScript-3178c6.svg?style=flat-square&logo=typescript&labelColor=444)
![Built for The Web](https://img.shields.io/badge/Built_for-The_Web-e34f26.svg?style=flat-square&logo=HTML5&labelColor=444)
![Runs Everywhere](https://img.shields.io/badge/Runs-Everywhere-f7df1e.svg?style=flat-square&logo=javascript&labelColor=444)

</div>

## Key Features

- **Fluent & Declarative APIs**: Use intuitive `html`, `dom`, and `svg` builders to declare elements and markup
- **Type-safe by design**: Comprehensive TypeScript support ensures strict types and intelligent autocompletion
- **Concurrent rendering**: Renders asynchronously for smoother interactions and declarative async operations
- **Native alignment**: 1:1 mapping to DOM propertiesâ€”no abstractions, just browser standards
- **Explicit contracts**: Props, children, and refs are strictly separated (no runtime ambiguity)
- **SSR-ready**: Render declarations, elements, and components to static markup
  - Pair with [Knyt Luthier](https://knyt.dev/s/luthier) for rendering closed declarative shadow roots
- **Stateless**: The DOM is the source of truth; no virtual DOM or reconciliation
- **Lightweight**: Minimal dependencies, minimal footprint
- **htmx-compatible**: First-class support for [htmx](https://htmx.org/) attributes

## Documentation

Documentation is available at [knyt.dev](https://knyt.dev), including a [guide on declarative rendering](https://knyt.dev/guide/declarative-rendering).

## Usage

The primary exports of this package are Declaration Builders and Declaration Processors. Declaration Builders are used to define the structure of the UI, while Declaration Processors handle the creation and manipulation of the actual DOM or markup. Together, these tools enable you to efficiently describe and render user interfaces.

- **Declaration Builders**: `html`, `dom`, and `svg` for creating declarations
- **Declaration Processors**: `build`, `update`, and `render` for DOM manipulation and rendering

In the example below, the `dom` builder is used to create an element declaration for a heading, and the `update` processor is used to update the `root` element to display the heading.

```ts
import { dom, update } from "@knyt/weaver";

const heading = dom.h1.$("Hello, world!");
const root = document.getElementById("app")!;

update(root, heading);
```

For cases that require more functionality, such as encapsulated behavior or custom lifecycle management, you can define custom web components using [Knyt Luthier](https://knyt.dev/s/luthier).

## Install

You can use this package by installing either the [Knyt Toolkit](https://knyt.dev/s/toolkit) or this standalone package.

_Knyt Toolkit:_

```sh
npm install knyt
```

_Standalone:_

```sh
npm install @knyt/weaver
```

## Updates

See the [CHANGELOG](./CHANGELOG.md) for a list of changes.

## License

This package is licensed under the [BSD 3-Clause License](./LICENSE).

## Open-Source Initiative

This project is supported by [Sables Applications LLC](https://sables.app) as part of its open-source initiatives.
