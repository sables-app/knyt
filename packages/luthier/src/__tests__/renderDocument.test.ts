/// <reference types="bun-types" />

import { dom, html, render } from "@knyt/weaver";
import { expect, test } from "bun:test";

import { define } from "../define/mod.ts";

const Head = define.view<{ title?: string; description?: string }>(
  ({ title, description }) => {
    return html`
      <head>
        <meta charset="UTF-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1.0" />
        <title>${title}</title>
        <meta name="description" content="${description}" />
        <link href="../styles/.tailwind.css" rel="stylesheet" />
        <script src="../scripts/main.ts" type="module"></script>
      </head>
    `;
  },
);

const MyHeading = define.element("my-heading-2239cacf90bf", {
  options: {
    shadowRoot: { mode: "closed" },
  },
  properties: {
    title: define.property().string().attribute("title"),
  },
  lifecycle: (host) => {
    return () => dom.h3.$(host.title);
  },
});

const documentDeclaration = html`<!doctype html>
  <html>
    ${Head().title("Hello World").description("This is a test.")}
    <body>
      <main>
        <h1>Hello World</h1>
        ${html.p.$("This is a test.")} ${MyHeading().title("Hello World")}
      </main>
    </body>
  </html>`;

test("document rendering", async () => {
  const htmlText = await render(documentDeclaration, {
    reactiveElementTimeout: 100,
    disableRenderToString: false,
  });

  expect(htmlText).toMatchSnapshot("document rendering");
});
