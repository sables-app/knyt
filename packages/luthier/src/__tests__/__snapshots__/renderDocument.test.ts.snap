// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`document rendering: document rendering 1`] = `
"<!doctype html>
  <html>
    
      <head>
        <meta charset="UTF-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1.0" />
        <title>Hello World</title>
        <meta name="description" content="This is a test." />
        <link href="../styles/.tailwind.css" rel="stylesheet" />
        <script src="../scripts/main.ts" type="module"></script>
      </head>
    
    <body>
      <main>
        <h1>Hello World</h1>
        <p>This is a test.</p> <my-heading title="Hello World"><template shadowrootmode="closed"><h3>Hello World</h3></template></my-heading>
      </main>
    </body>
  </html>"
`;
